File: core/tmp/Java/src/com/dbc/algorithm/Sort/KthLargestElementInAnArray.java
package com.dbc.algorithm.Sort;

public class KthLargestElementInAnArray {
    public static int quickFind(int[] nums, int left, int right, int k) {
        int l = left, r = right;
        int mid = nums[left];
        while (l < r) {
            while (l < r && nums[r] >= mid) {
                r--;
            }
            if (l < r) {
                nums[l++] = nums[r];
            }
            while (l < r && nums[l] <= mid) {
                l++;
            }
            if (l < r) {
                nums[r--] = nums[l];
            }
        }
        nums[l] = mid;
        if (l + 1 == k) {
            return mid;
        } else if (l + 1 > k) {
            return quickFind(nums, left, l - 1, k);
        } else {
            return quickFind(nums, l + 1, right, k);
        }
    }

    public static int findKthLargestQuickFind(int[] nums, int k) {
        return quickFind(nums, 0, nums.length - 1, nums.length + 1 - k);
    }

    public static void swap(int[] nums, int index1, int index2) {
        int temp = nums[index1];
        nums[index1] = nums[index2];
        nums[index2] = temp;
    }

    public static void heapify(int[] nums, int i, int length) {
        int left = 2 * i + 1;
        int right = 2 * i + 2;
        int largest = i;
        if (left < length && nums[largest] < nums[left]) {
            largest = left;
        }
        if (right < length && nums[largest] < nums[right]) {
            largest = right;
        }

        if (largest != i) {
            swap(nums, i, largest);
            heapify(nums, largest, length);
        }
    }

    public static int findKthLargestHeap(int[] nums, int k) {
        for (int i = Math.floorDiv(nums.length, 2); i > 0; i--) {
            heapify(nums, i - 1, nums.length);
        }

        for (int i = 0; i < k - 1; i++) {
            swap(nums, 0, nums.length - 1 - i);
            heapify(nums, 0, nums.length - 1 - i);
        }
        return nums[0];
    }
}

File: core/tmp/Python/sort/kth_largest_element_in_an_array.py
from typing import List


def quick_find(nums: List[int], left: int, right: int, k: int):
    l = left
    r = right
    mid = nums[left]
    while l < r:
        while l < r and nums[r] >= mid:
            r -= 1
        if l < r:
            nums[l] = nums[r]
            l += 1
        while l < r and nums[l] <= mid:
            l += 1
        if l < r:
            nums[r] = nums[l]
            r -= 1
    nums[l] = mid
    if l + 1 == k:
        return mid
    elif l + 1 > k:
        return quick_find(nums, left, l - 1, k)
    else:
        return quick_find(nums, l + 1, right, k)


def findKthLargestQuick(nums: List[int], k: int) -> int:
    return quick_find(nums, 0, len(nums) - 1, len(nums) - k + 1)


def swap(nums: List[int], index1: int, index2: int):
    temp = nums[index1]
    nums[index1] = nums[index2]
    nums[index2] = temp


def heapify(nums: List[int], i: int, length: int):
    left = 2 * i + 1
    right = 2 * i + 2
    largest = i
    if left < length and nums[largest] < nums[left]:
        largest = left
    if right < length and nums[largest] < nums[right]:
        largest = right

    if i != largest:
        swap(nums, largest, i)
        heapify(nums, largest, length)


def findKthLargestHeap(nums: List[int], k: int) -> int:
    length = len(nums)
    for i in range(length // 2, 0, -1):
        heapify(nums, i - 1, length)

    for i in range(k - 1):
        swap(nums, 0, length - 1 - i)
        heapify(nums, 0, length - 1 - i)

    return nums[0]


if __name__ == "__main__":

    print(findKthLargestHeap(nums=[2,1], k=2))
